### PR TITLE
bpo-9303: Migrate sqlite3 module to _v2 API to enhance performance

### PR DESCRIPTION
**Pull request by palaviv**
Use the _v2 API when possible.

### CODE REVIEW

**Review by berkerpeksag — COMMENTED** [review]
I think we can also avoid calling `sqlite3_reset` in `_pysqlite_seterror` if the v2 API is available.

**Comment by berkerpeksag** [comment]
I haven't checked yet, but I also think that the check for `SQLITE_SCHEMA` can also be eliminated if the v2 API is available: https://github.com/python/cpython/blob/c92e6047bde01bcb846ad1fcf4126d3001e30f4b/Modules/_sqlite/cursor.c#L553

File: Modules/_sqlite/connection.c
Comment (berkerpeksag) [inline Modules/_sqlite/connection.c]: Style nitpick: `FOO|BAR` is more common than `FOO | BAR` in sqlite3 code base.

File: Modules/_sqlite/connection.c
Comment (berkerpeksag) [inline Modules/_sqlite/connection.c]: I think that there is an opportunity to simplify this whole `SQLITE_OPEN_URI` and `HAVE_SQLITE3_OPEN_V2` dance.

(The use of `Py_BEGIN_ALLOW_THREADS` is a bit weird in line 111 and line 120.)

**Comment by palaviv** [comment]
> I think we can also avoid calling sqlite3_reset in _pysqlite_seterror if the v2 API is available.

done

> I haven't checked yet, but I also think that the check for SQLITE_SCHEMA can also be eliminated if the v2 API is available: https://github.com/python/cpython/blob/c92e6047bde01bcb846ad1fcf4126d3001e30f4b/Modules/_sqlite/cursor.c#L553

I don't think we should do this because we can get `SQLITE_SCHEMA` if `SQLITE_MAX_SCHEMA_RETRY` tries to recompile will pass.

File: Modules/_sqlite/util.c
Comment (berkerpeksag) [inline Modules/_sqlite/util.c]: Nitpick: Please remove extra empty line.

File: Modules/_sqlite/connection.c
Comment (berkerpeksag) [inline Modules/_sqlite/connection.c]: Nitpick: `FOO|BAR` syntax instead of `FOO | BAR`.

Also, finish the line with a full stop and space:

```c
/* [...] NULL). */
```

**Review by berkerpeksag — COMMENTED** [review]
This looks pretty good to me. We just need an entry in `Misc/NEWS` (please add "Patch by Aviv Palivoda." too)

**Comment by berkerpeksag** [comment]
> I don't think we should do this because we can get SQLITE_SCHEMA if SQLITE_MAX_SCHEMA_RETRY tries to recompile will pass.

Right, we can still get `SQLITE_SCHEMA` from `sqlite3_step()` (in legacy API we need to call `sqlite3_reset()` to get it), but in this case `rc` is the return value of `sqlite3_reset()`. The latter call is not needed when we have the new API since `sqlite3_step()` calls `sqlite3_reset()` for us.

File: Modules/_sqlite/connection.c
Comment (palaviv) [reply Modules/_sqlite/connection.c]: done

File: Modules/_sqlite/util.c
Comment (palaviv) [reply Modules/_sqlite/util.c]: done

**Comment by palaviv** [comment]
I have removed the reset but I left the recompile because we should recompile before calling sqlite3_step again.

**Review by serhiy-storchaka — APPROVED** [review]

File: Modules/_sqlite/util.c
Comment (berkerpeksag) [inline Modules/_sqlite/util.c]: Shouldn't `>=` be `<` here?

Comment (serhiy-storchaka) [reply Modules/_sqlite/util.c]: Good catch!

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]
Fix an error found by @berkerpeksag.

File: Modules/_sqlite/util.c
Comment (palaviv) [reply Modules/_sqlite/util.c]: Fixed

**Review by berkerpeksag — APPROVED** [review]
Thanks, Aviv! I will wait for Serhiy's approval and merge this.

**Review by serhiy-storchaka — APPROVED** [review]

**Pull request merged**
